pipeline/trusted-resources/pkg/trustedtask/trustedtaskrun_test.go

```diff
@@ -0,0 +1,601 @@
+		TypeMeta:   trTypeMeta,
+		ObjectMeta: trObjectMeta,
+		Spec: v1beta1.TaskRunSpec{
+			TaskSpec: &ts.Spec,
+		},
+	}
+
+	apiTaskRun := TrustedTaskRun{TaskRun: tr}
+	taskRunPass := apiTaskRun.DeepCopyObject().(*TrustedTaskRun)
+	taskRunContentTempered := apiTaskRun.DeepCopyObject().(*TrustedTaskRun)
+	taskRunUnsigned := apiTaskRun.DeepCopyObject().(*TrustedTaskRun)
+
+	// Create registry server
+	s := httptest.NewServer(registry.New())
+	defer s.Close()
+	u, _ := url.Parse(s.URL)
+
+	// Push OCI bundle
+	dig, err := pushOCIImage(t, u, &ts)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	digTempered, err := pushOCIImage(t, u, &tsTempered)
+	t.Logf("Digest: %v", digTempered.String())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// OCI taskruns
+	otr := v1beta1.TaskRun{
+		TypeMeta:   trTypeMeta,
+		ObjectMeta: trObjectMeta,
+		Spec: v1beta1.TaskRunSpec{
+			TaskRef: &v1beta1.TaskRef{
+				Name:   "ts",
+				Bundle: u.Host + "/task/" + ts.Name,
+			},
+		},
+	}
+
+	OCITaskRun := TrustedTaskRun{TaskRun: otr}
+	taskRunPassOCIBundle := OCITaskRun.DeepCopyObject().(*TrustedTaskRun)
+	taskRunPassOCIBundleContentTempered := OCITaskRun.DeepCopyObject().(*TrustedTaskRun)
+	taskRunPassOCIBundleUnsigned := OCITaskRun.DeepCopyObject().(*TrustedTaskRun)
+
+	// Local taskref taskruns
+	ltr := v1beta1.TaskRun{
+		TypeMeta:   trTypeMeta,
+		ObjectMeta: trObjectMeta,
+		Spec: v1beta1.TaskRunSpec{
+			TaskRef: &v1beta1.TaskRef{
+				Name: "test-task",
+			},
+		},
+	}
+
+	localRefTaskRun := TrustedTaskRun{TaskRun: ltr}
+
+	taskRunPassTaskRef := localRefTaskRun.DeepCopyObject().(*TrustedTaskRun)
+	taskRunPassTaskRefContentTempered := localRefTaskRun.DeepCopyObject().(*TrustedTaskRun)
+	taskRunPassTaskRefUnsigned := localRefTaskRun.DeepCopyObject().(*TrustedTaskRun)
+
+	// Sign task
+	trList := []*TrustedTaskRun{
+		taskRunPass,
+		taskRunContentTempered,
+		taskRunPassOCIBundle,
+		taskRunPassOCIBundleContentTempered,
+		taskRunPassTaskRef,
+		taskRunPassTaskRefContentTempered,
+	}
+
+	for _, tr := range trList {
+		if tr.Spec.TaskSpec != nil {
+			signatureEncoding := signTaskSpec(t, signer, tr.Spec.TaskSpec)
+			tr.Annotations["tekton.dev/signature"] = signatureEncoding
+		}
+		if tr.Spec.TaskRef != nil {
+			if tr.Spec.TaskRef.Bundle != "" {
+				signatureEncoding := signOCIBundle(t, signer, dig)
+				tr.Annotations["tekton.dev/signature"] = signatureEncoding
+
+			} else if ts.Name != "" {
+				ts, err := tektonClient.TektonV1beta1().Tasks(tr.Namespace).Get(ctx, tr.Spec.TaskRef.Name, metav1.GetOptions{})
+				if err != nil {
+					t.Fatalf("Unexpected err %v", err)
+				}
+				signatureEncoding := signTaskSpec(t, signer, &ts.Spec)
+				tr.Annotations["tekton.dev/signature"] = signatureEncoding
+			}
+		}
+	}
+
+	// tamper taskSpec data after signed
+	taskRunContentTempered.Spec.TaskSpec = &tsTempered.Spec
+	taskRunPassOCIBundleContentTempered.Spec.TaskRef.Bundle = u.Host + "/task/" + tsTempered.Name
+	taskRunPassTaskRefContentTempered.Spec.TaskRef.Name = tsTempered.Name
```

Comment:
This is a good example over where a table-driven test may not make as much sense as individual tests.

e.g. because we're doing different test / environment set up for each of Embedded TaskSpec / TaskRef / Task Bundle makes it difficult to follow what is happening. A reader needs to keep switching contexts for each type, and even authors, need to add logic like the signing code above to recategorize each type to do specific things.

I'd recommend breaking this up into separate tests for each type.